src/registry-auth-locator/config.ts

```diff
@@ -0,0 +1,37 @@
+import os from "os";
+import path from "path";
+import { DockerConfig } from "./types";
+import { existsSync, promises as fs } from "fs";
+
+const dockerConfigLocation = process.env.DOCKER_CONFIG || `${os.homedir()}/.docker`;
+
+const dockerConfigFile = path.resolve(dockerConfigLocation, "config.json");
+
+/**
+ * Reads the docker config file from environment variable and returns the {@link DockerConfig}.
+ *
+ * Some CI environments (e.g. GitLab) encourages configuring docker using the environment variable
+ * DOCKER_AUTH_CONFIG. If it exists, it will be used when there is no default config file.
+ */
+const readDockerConfigFromEnv = (): DockerConfig => {
+  const dockerAuthConfig = process.env.DOCKER_AUTH_CONFIG;
+  if (!dockerAuthConfig) {
+    return {};
+  }
+  return JSON.parse(dockerAuthConfig);
+};
+
+export const readDockerConfig = async (): Promise<DockerConfig> => {
+  if (!existsSync(dockerConfigFile)) {
+    return readDockerConfigFromEnv();
```

Comment:
Could this be refactored to show the 3 different paths at a high level? Something like:

```javascript
export const readDockerConfig = async () => {
  if (existsSync(dockerConfigFile)) {
    return fromDockerConfigFile();
  } else if (process.env.DOCKER_AUTH_CONFIG !== undefined) {
    return fromDockerAuthConfigEnv();
  } else {
    return {};
  }
};
```

Reply:
Check out the latest solution and see if looks better.